Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt

```diff
@@ -57,33 +67,77 @@ class CheckInsFragment : Fragment(R.layout.trace_location_attendee_checkins_frag
-            ) { position, direction ->
-                val checkInsItem = checkInsAdapter.data[position]
-                if (checkInsItem.isSwipeable()) {
-                    checkInsItem.onSwipe(position, direction)
-                }
-            }
+        viewModel.checkins.observe2(this) { items ->
+            updateViews(items)
         }
 
-        viewModel.checkins.observe2(this) {
-            checkInsAdapter.update(it)
-            binding.apply {
-                checkInsList.isGone = it.isEmpty()
-                emptyListInfoContainer.isGone = it.isNotEmpty()
-            }
+        viewModel.events.observe2(this) {
+            onNavigationEvent(it)
+        }
+
+        viewModel.errorEvent.observe2(this) {
+            val errorForHumans = it.tryHumanReadableError(requireContext())
+            Toast.makeText(requireContext(), errorForHumans.description, Toast.LENGTH_LONG).show()
+        }
+    }
+
+    private fun onNavigationEvent(event: CheckInEvent?) {
+        when (event) {
+            is CheckInEvent.ConfirmCheckIn -> doNavigate(
+                CheckInsFragmentDirections.actionCheckInsFragmentToConfirmCheckInFragment(
+                    verifiedTraceLocation = event.verifiedTraceLocation,
+                    editCheckInId = 0,
+                )
+            )
+
+            is CheckInEvent.ConfirmSwipeItem -> showRemovalConfirmation(event.checkIn, event.position)
+
+            is CheckInEvent.ConfirmRemoveItem -> showRemovalConfirmation(event.checkIn, null)
+
+            is CheckInEvent.ConfirmRemoveAll -> showRemovalConfirmation(null, null)
+
+            is CheckInEvent.EditCheckIn -> doNavigate(
+                CheckInsFragmentDirections.actionCheckInsFragmentToConfirmCheckInFragment(
+                    verifiedTraceLocation = null,
+                    editCheckInId = event.checkInId,
+                )
+            )
+
+            is CheckInEvent.ShowInformation -> Toast.makeText(
+                requireContext(),
+                "TODO ¯\\_(ツ)_/¯",
+                Toast.LENGTH_SHORT
+            ).show()
+
+            is CheckInEvent.OpenDeviceSettings -> openDeviceSettings()
+        }
+    }
+
+    private fun updateViews(items: List<CheckInsItem>) {
+        checkInsAdapter.update(items)
+        binding.apply {
+            val index = items.indexOfFirst { it is CameraPermissionVH.Item }
+            swipeExcludedPositions.replaceAll(listOf(index))
+
+            scanCheckinQrcodeFab.isGone = items.any { it is CameraPermissionVH.Item }
+            emptyListInfoContainer.isGone = items.isNotEmpty()
+            checkInsList.isGone = items.isEmpty()
         }
+    }
 
+    private fun openDeviceSettings() {
+        startActivity(
+            Intent(
+                Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
+                Uri.parse("package:" + BuildConfig.APPLICATION_ID)
+            )
+        )
```

Comment:
Please catch the ActivityNotFoundException here.

Reply:
Do you expect that any App won't have a settings screen in Android 🤔? If yes what do you want to show to the user when he clicks on the card and nothing happens?

Reply:
> Do you expect that any App won't have a settings screen in Android thinking? 

Yes, I sadly have come to the conclusion that any activity you launch, that is outside of your app, will eventually cause an `ActivityNotFoundException` on some device.
It will be rare, but we shouldn't crash the app. Showing the exception in a toast would be fine for me.